File: packages/sdk/src/clients.ts
import { ethers, type Contract, type ContractRunner, type ContractTransactionReceipt } from "ethers";
import { RON_ABI, CAPABILITY_REGISTRY_ABI } from "./abis.js";
import type { Capability, Completion, CreditScore, ScoreDetails, SelfAttestScore, ScoreType, TwoSidedScore } from "./types.js";

/** Extract an event argument from a receipt (v0.1.2): ids must come from the
 *  emitted event, never from supply counters (which diverge after burns). */
function parseEventId(
  receipt: ContractTransactionReceipt | null,
  contract: Contract,
  eventName: string,
  argName: string,
): bigint | null {
  if (!receipt) return null;
  for (const log of receipt.logs) {
    try {
      const parsed = contract.interface.parseLog(log);
      if (parsed && parsed.name === eventName) {
        const value = parsed.args.getValue(argName) ?? parsed.args[0];
        return BigInt(value);
      }
    } catch {
      // log emitted by another contract in the same tx — ignore
    }
  }
  return null;
}

export class ReputationOracleNetworkClient {
  private readonly c: Contract;
  constructor(address: string, runner?: ContractRunner) {
    this.c = new ethers.Contract(address, RON_ABI, runner);
  }
  get address(): string {
    return this.c.target as string;
  }
  completionCount(agent: string): Promise<bigint> {
    return this.c.completionCount(agent) as Promise<bigint>;
  }
  /** v0.2: number of receipt-confirmed (two-sided) completions. */
  confirmedCount(agent: string): Promise<bigint> {
    return this.c.confirmedCount(agent) as Promise<bigint>;
  }
  disputeCount(agent: string): Promise<bigint> {
    return this.c.disputeCount(agent) as Promise<bigint>;
  }

  async attestCompletion(taskType: string, resultCID: string): Promise<{ completionId: bigint; receipt: ContractTransactionReceipt | null }> {
    const tx = await this.c.attestCompletion(ethers.id(taskType), resultCID);
    const receipt = (await tx.wait()) ?? null;
    // v0.1.2: read the id from the emitted event. Reading `nextCompletionId()`
    // after the tx breaks under concurrency (two attestations in flight).
    const completionId =
      parseEventId(receipt, this.c, "SelfAttested", "completionId") ?? (await this.c.nextCompletionId() as bigint);
    return { completionId, receipt };
  }
  /** v0.2: countersign a completion as the independent requester (two-sided). */
  async attestReceipt(completionId: number | bigint, receiptCID: string): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.attestReceipt(completionId, receiptCID)).wait()) ?? null;
  }
  /** v0.2: withdraw a previously-given receipt (counterparty only). */
  async revokeReceipt(completionId: number | bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.revokeReceipt(completionId)).wait()) ?? null;
  }
  async challengeCompletion(completionId: number | bigint, evidenceCID: string, bond: bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.challengeCompletion(completionId, evidenceCID, { value: bond })).wait()) ?? null;
  }
  /** v0.2: agent rebuts a challenge within the challenge window. */
  async contestChallenge(completionId: number | bigint, rebuttalCID: string): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.contestChallenge(completionId, rebuttalCID)).wait()) ?? null;
  }
  /** v0.2: finalize an uncontested challenge after the window (uplheld optimistically). */
  async finalizeChallenge(completionId: number | bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.finalizeChallenge(completionId)).wait()) ?? null;
  }
  async resolveChallenge(completionId: number | bigint, upheld: boolean): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.resolveChallenge(completionId, upheld)).wait()) ?? null;
  }
  async getSelfAttestScore(agent: string): Promise<SelfAttestScore> {
    const r = (await this.c.getSelfAttestScore(agent)) as [bigint, bigint, bigint];
    return { completions: r[0], disputes: r[1], score: r[2] };
  }
  /** v0.2: score based on receipt-confirmed completions. Prefer this over
   *  `getSelfAttestScore` when ranking agents. */
  async getTwoSidedScore(agent: string): Promise<TwoSidedScore> {
    const r = (await this.c.getTwoSidedScore(agent)) as [bigint, bigint, bigint, bigint, number];
    return { confirmed: r[0], disputes: r[1], score: r[2], lastActivity: r[3], decayBps: Number(r[4]) };
  }
  /** v0.3: diversity-adjusted credit score (distinct counterparties - disputes). */
  async getCreditScore(agent: string): Promise<CreditScore> {
    const r = (await this.c.getCreditScore(agent)) as [bigint, bigint, bigint, bigint, number];
    return { distinctCounterparties: r[0], disputes: r[1], score: r[2], lastActivity: r[3], decayBps: Number(r[4]) };
  }
  /** v0.3: number of distinct counterparties that have confirmed this agent. */
  distinctCounterparties(agent: string): Promise<bigint> {
    return this.c.distinctCounterparties(agent) as Promise<bigint>;
  }
  /** v0.3: per-(agent, counterparty) confirmation count. */
  counterpartyConfirmations(agent: string, counterparty: string): Promise<bigint> {
    return this.c.counterpartyConfirmations(agent, counterparty) as Promise<bigint>;
  }
  /** v0.3: attestation cooldown in seconds (0 = off). */
  attestCooldown(): Promise<bigint> {
    return this.c.attestCooldown() as Promise<bigint>;
  }
  /** v0.3: set the attestation cooldown (owner/Timelock only). */
  async setAttestCooldown(cooldown: number | bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.setAttestCooldown(cooldown)).wait()) ?? null;
  }
  /** v0.3: pause protocol actions (owner/Timelock only). */
  async pause(): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.pause()).wait()) ?? null;
  }
  /** v0.3: unpause protocol actions (owner/Timelock only). */
  async unpause(): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.unpause()).wait()) ?? null;
  }
  /** v0.3: is the contract paused? */
  paused(): Promise<boolean> {
    return this.c.paused() as Promise<boolean>;
  }
  /** Best available ranking signal: credit (v0.3) -> two-sided (v0.2) -> self-attest. */
  async getRankingScore(agent: string): Promise<{ score: bigint; completions: bigint; disputes: bigint; scoreType: ScoreType }> {
    try {
      const s = await this.getCreditScore(agent);
      return { score: s.score, completions: s.distinctCounterparties, disputes: s.disputes, scoreType: "credit" };
    } catch {
      /* pre-v0.3 */
    }
    try {
      const s = await this.getTwoSidedScore(agent);
      return { score: s.score, completions: s.confirmed, disputes: s.disputes, scoreType: "two-sided" };
    } catch {
      /* pre-v0.2 */
    }
    const s = await this.getSelfAttestScore(agent);
    return { score: s.score, completions: s.completions, disputes: s.disputes, scoreType: "self-attest" };
  }
  async getScoreDetails(agent: string): Promise<ScoreDetails> {
    const r = (await this.c.getScoreDetails(agent)) as [bigint, bigint, bigint, bigint, number];
    return { completions: r[0], disputes: r[1], score: r[2], lastActivity: r[3], decayBps: Number(r[4]) };
  }
  async getCompletion(id: number | bigint): Promise<Completion> {
    const r = (await this.c.getCompletion(id)) as [string, string, string, bigint, boolean, boolean, string, bigint];
    return {
      agent: r[0],
      taskType: r[1],
      resultCID: r[2],
      timestamp: r[3],
      challenged: r[4],
      disputed: r[5],
      counterparty: r[6] ?? ethers.ZeroAddress,
      receiptTimestamp: r[7] ?? 0n,
    };
  }
  /** v0.2: the receipt evidence CID for a completion (empty string if none). */
  receiptCID(completionId: number | bigint): Promise<string> {
    return this.c.receiptCID(completionId) as Promise<string>;
  }
  async withdrawEthPool(to: string, amount: bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.withdrawEthPool(to, amount)).wait()) ?? null;
  }
  challengeBond(): Promise<bigint> {
    return this.c.CHALLENGE_BOND() as Promise<bigint>;
  }
  /** v0.2: seconds the agent has to contest a challenge. */
  challengeWindow(): Promise<bigint> {
    return this.c.CHALLENGE_WINDOW() as Promise<bigint>;
  }

  // Basic agent identity (Step 7)
  lastActivity(agent: string): Promise<bigint> {
    return this.c.lastActivity(agent) as Promise<bigint>;
  }
  agentMetadataCID(agent: string): Promise<string> {
    return this.c.agentMetadataCID(agent) as Promise<string>;
  }
  async registerAgent(metadataCID: string): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.registerAgent(metadataCID)).wait()) ?? null;
  }
  getAgentMetadata(agent: string): Promise<string> {
    return this.c.getAgentMetadata(agent) as Promise<string>;
  }
}

export class CapabilityRegistryClient {
  private readonly c: Contract;
  constructor(address: string, runner?: ContractRunner) {
    this.c = new ethers.Contract(address, CAPABILITY_REGISTRY_ABI, runner);
  }
  get address(): string {
    return this.c.target as string;
  }
  totalSupply(): Promise<bigint> {
    return this.c.totalSupply() as Promise<bigint>;
  }
  tokenByIndex(index: number | bigint): Promise<bigint> {
    return this.c.tokenByIndex(index) as Promise<bigint>;
  }
  ownerOf(id: number | bigint): Promise<string> {
    return this.c.ownerOf(id) as Promise<string>;
  }
  async getCapability(id: number | bigint): Promise<Capability> {
    const raw = (await this.c.getCapability(id)) as [string, bigint, string, string, boolean, boolean];
    return {
      creator: raw[0],
      bond: raw[1],
      capabilityType: raw[2],
      metadataCID: raw[3],
      certified: raw[4],
      slashed: raw[5],
    };
  }
  async registerCapabilityEth(capabilityType: string, metadataCID: string, bondWei: bigint): Promise<{ capabilityId: bigint; receipt: ContractTransactionReceipt | null }> {
    const tx = await this.c.registerCapabilityEth(ethers.id(capabilityType), metadataCID, { value: bondWei });
    const receipt = (await tx.wait()) ?? null;
    // v0.1.2: derive the id from the emitted event. `totalSupply()` returns the
    // wrong id once any capability has been withdrawn (burn decrements supply
    // but ids keep incrementing) — which made the backend certify the wrong NFT.
    const capabilityId =
      parseEventId(receipt, this.c, "CapabilityRegistered", "capabilityId") ??
      (await this.c.totalSupply() as bigint);
    return { capabilityId, receipt };
  }
  async certifyCapability(id: number | bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.certifyCapability(id)).wait()) ?? null;
  }
  async slashCapability(id: number | bigint, penalty: bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.slashCapability(id, penalty)).wait()) ?? null;
  }
  async withdrawBond(id: number | bigint): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.withdrawBond(id)).wait()) ?? null;
  }

  async getCapabilitiesByType(capabilityType: string): Promise<bigint[]> {
    const ids = (await this.c.getCapabilitiesByType(ethers.id(capabilityType))) as bigint[];
    return ids;
  }

  /** F10: number of live capabilities of a type (page without loading the array). */
  countCapabilitiesByType(capabilityType: string): Promise<bigint> {
    return this.c.countCapabilitiesByType(ethers.id(capabilityType)) as Promise<bigint>;
  }

  /** F10: paginated discovery view (v0.2 contracts). */
  async getCapabilitiesByTypePaged(
    capabilityType: string,
    offset: number | bigint,
    limit: number | bigint,
  ): Promise<bigint[]> {
    return (await this.c.getCapabilitiesByTypePaged(ethers.id(capabilityType), offset, limit)) as bigint[];
  }

  /** v0.3: pause registry actions (owner/Timelock only; creator exits stay open). */
  async pause(): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.pause()).wait()) ?? null;
  }
  /** v0.3: unpause registry actions (owner/Timelock only). */
  async unpause(): Promise<ContractTransactionReceipt | null> {
    return (await (await this.c.unpause()).wait()) ?? null;
  }
  /** v0.3: is the registry paused? */
  paused(): Promise<boolean> {
    return this.c.paused() as Promise<boolean>;
  }
}

export type DiscoveryItem = {
  agentAddress: string;
  capabilityId: bigint;
  capabilityType: string;
  certified: boolean;
  slashed: boolean;
  bond: bigint;
  metadataCID: string;
  completions: bigint;
  disputes: bigint;
  score: bigint;
  /** Which signal the ranking score used (credit > two-sided > self-attest). */
  scoreType: ScoreType;
};

export async function discover(
  registry: CapabilityRegistryClient,
  ron: ReputationOracleNetworkClient,
  capabilityType = "LoRA",
  minScore = 0n as bigint | number,
): Promise<DiscoveryItem[]> {
  const min = typeof minScore === "bigint" ? minScore : BigInt(minScore);
  let ids: bigint[];
  try {
    ids = await registry.getCapabilitiesByType(capabilityType);
  } catch {
    const total = await registry.totalSupply();
    ids = [];
    for (let i = 0n; i < total; i++) ids.push(await registry.tokenByIndex(i));
  }
  const out: DiscoveryItem[] = [];
  for (const capId of ids) {
    // v0.1.2: a single stale/unreadable id (e.g. pre-fix index pollution, or an
    // id burned on another deployment) must never break the whole discovery call.
    let cap;
    try {
      cap = await registry.getCapability(capId);
    } catch {
      continue;
    }
    if (cap.capabilityType.toLowerCase() !== ethers.id(capabilityType).toLowerCase()) continue;
    if (!cap.certified || cap.slashed) continue;
    // v0.2: rank on the two-sided (receipt-confirmed) score where the contract
    // supports it, falling back to the legacy self-attest score otherwise.
    const s = await ron.getRankingScore(cap.creator);
    if (s.score < min) continue;
    out.push({
      agentAddress: cap.creator,
      capabilityId: capId,
      capabilityType,
      certified: cap.certified,
      slashed: cap.slashed,
      bond: cap.bond,
      metadataCID: cap.metadataCID,
      completions: s.completions,
      disputes: s.disputes,
      score: s.score,
      scoreType: s.scoreType,
    });
  }
  out.sort((a, b) => (b.score > a.score ? 1 : b.score < a.score ? -1 : 0));
  return out;
}

/** Load deployments.json written by scripts/deploy-*.ts. */
export async function loadDeployment(path: string) {
  const fs = await import("node:fs/promises");
  const raw = await fs.readFile(path, "utf8");
  return JSON.parse(raw) as import("./types.js").Deployment;
}

File: packages/python-sdk/taop/clients.py
"""Typed contract clients for TAOP MVP — mirrors @taopp/sdk (TypeScript)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Optional, Union

from eth_account import Account
from eth_utils import to_hex
from web3 import Web3
from web3.contract import AsyncContract, Contract
from web3.middleware import ExtraDataToPOAMiddleware

from .abis import CAPABILITY_REGISTRY_ABI, RON_ABI
from .types import (
    AgentScore,
    Capability,
    Completion,
    Deployment,
    SelfAttestScore,
    TwoSidedScore,
    LORA_CAPABILITY_TYPE,
)


def _keccak(s: str) -> bytes:
    return Web3.keccak(text=s)


def _load_deployment(path: Union[str, Path]) -> Deployment:
    data = json.loads(Path(path).read_text())
    # `token` / `validatorStake` only exist in local-deploy output; the live
    # deployments.json is addresses-only (v0.1.2 removed key material from it).
    return Deployment(
        chain_id=data["chainId"],
        token=data.get("token", "0x0000000000000000000000000000000000000000"),
        ron=data["ron"],
        registry=data["registry"],
        validator=data["validator"],
        agent_a=data["agentA"],
        agent_pk=data.get("agentAPk", ""),
        validator_stake=data.get("validatorStake", "0"),
        network=data.get("network", ""),
        deployed_at=data.get("deployedAt", ""),
    )


def _event_id_from_receipt(contract, receipt, event_name: str, arg_name: str) -> Optional[int]:
    """v0.1.2: derive ids from the emitted event, never from supply counters
    (totalSupply()/nextCompletionId() diverge once capabilities are burned)."""
    if receipt is None:
        return None
    try:
        event = getattr(contract.events, event_name)
    except Exception:
        return None
    for log in receipt.get("logs", []) or []:
        try:
            parsed = event().process_log(log)
            return int(parsed["args"][arg_name])
        except Exception:
            continue
    return None


class ReputationOracleNetworkClient:
    """Client for the ReputationOracleNetwork (Credit Bureau) contract."""

    def __init__(self, address: str, w3: Web3, account: Optional[Account] = None):
        self.w3 = w3
        self.account = account
        self.contract = w3.eth.contract(address=address, abi=RON_ABI)

    @property
    def address(self) -> str:
        return self.contract.address

    def get_agent_score(self, agent: str) -> AgentScore:
        """Removed in v0.1.2: `getAgentScore` was dormant v2 ABI that reverts
        against the deployed v1 contracts. Use `get_self_attest_score`."""
        raise NotImplementedError(
            "get_agent_score was removed in v0.1.2 (dormant v2 ABI). Use get_self_attest_score()."
        )

    def get_self_attest_score(self, agent: str) -> SelfAttestScore:
        r = self.contract.functions.getSelfAttestScore(agent).call()
        return SelfAttestScore(completions=r[0], disputes=r[1], score=r[2])

    def get_two_sided_score(self, agent: str) -> TwoSidedScore:
        """v0.2: score based on receipt-confirmed completions (prefer for ranking)."""
        r = self.contract.functions.getTwoSidedScore(agent).call()
        return TwoSidedScore(confirmed=r[0], disputes=r[1], score=r[2], last_activity=r[3], decay_bps=r[4])

    def get_ranking_score(self, agent: str) -> tuple[int, int, int, str]:
        """v0.2: two-sided score where the contract supports it, else self-attest.
        Returns (score, count, disputes, score_type)."""
        try:
            s = self.get_two_sided_score(agent)
            return s.score, s.confirmed, s.disputes, "two-sided"
        except Exception:
            s = self.get_self_attest_score(agent)
            return s.score, s.completions, s.disputes, "self-attest"

    def get_completion(self, completion_id: int) -> Completion:
        r = self.contract.functions.getCompletion(completion_id).call()
        return Completion(
            agent=r[0], task_type=r[1], result_cid=r[2], timestamp=r[3],
            challenged=r[4], disputed=r[5],
            counterparty=r[6] if len(r) > 6 else "0x0000000000000000000000000000000000000000",
            receipt_timestamp=r[7] if len(r) > 7 else 0,
        )

    def completion_count(self, agent: str) -> int:
        return self.contract.functions.completionCount(agent).call()

    def confirmed_count(self, agent: str) -> int:
        """v0.2: number of receipt-confirmed completions."""
        return self.contract.functions.confirmedCount(agent).call()

    def receipt_cid(self, completion_id: int) -> str:
        """v0.2: the requester's receipt evidence CID (empty if none)."""
        return self.contract.functions.receiptCID(completion_id).call()

    def challenge_bond(self) -> int:
        return self.contract.functions.CHALLENGE_BOND().call()

    def challenge_window(self) -> int:
        """v0.2: seconds the agent has to contest a challenge."""
        return self.contract.functions.CHALLENGE_WINDOW().call()

    def _send_tx(self, tx):
        if self.account is None:
            raise ValueError("No account set — cannot send transactions")
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        return self.w3.eth.wait_for_transaction_receipt(tx_hash)

    # Basic agent identity (Step 7)
    def last_activity(self, agent: str) -> int:
        return self.contract.functions.lastActivity(agent).call()

    def agent_metadata_cid(self, agent: str) -> str:
        return self.contract.functions.agentMetadataCID(agent).call()

    def register_agent(self, metadata_cid: str) -> dict:
        if self.account is None:
            raise ValueError("No account set")
        fn = self.contract.functions.registerAgent(metadata_cid)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 100_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        receipt = self._send_tx(tx)
        return {"receipt": receipt}

    def attest_completion(self, task_type: str, result_cid: str) -> dict:
        """Self-attest a completion. Returns {'completionId': int, 'receipt': dict}."""
        fn = self.contract.functions.attestCompletion(_keccak(task_type), result_cid)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        receipt = self._send_tx(tx)
        # v0.1.2: read the id from the emitted event (concurrency-safe).
        completion_id = _event_id_from_receipt(self.contract, receipt, "SelfAttested", "completionId")
        if completion_id is None:
            completion_id = self.contract.functions.nextCompletionId().call()
        return {"completionId": completion_id, "receipt": receipt}

    def challenge_completion(self, completion_id: int, evidence_cid: str, bond_wei: int) -> dict:
        bond = self.challenge_bond() if bond_wei is None else bond_wei
        fn = self.contract.functions.challengeCompletion(completion_id, evidence_cid)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "value": bond,
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)

    def resolve_challenge(self, completion_id: int, upheld: bool) -> dict:
        fn = self.contract.functions.resolveChallenge(completion_id, upheld)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)

    # --- v0.2: two-sided attestation + optimistic challenge window ---

    def attest_receipt(self, completion_id: int, receipt_cid: str) -> dict:
        """Counterparty countersigns a completion (turns a self-report two-sided)."""
        fn = self.contract.functions.attestReceipt(completion_id, receipt_cid)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)

    def revoke_receipt(self, completion_id: int) -> dict:
        """Counterparty withdraws a receipt."""
        fn = self.contract.functions.revokeReceipt(completion_id)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)

    def contest_challenge(self, completion_id: int, rebuttal_cid: str) -> dict:
        """The agent rebuts a challenge within the challenge window."""
        fn = self.contract.functions.contestChallenge(completion_id, rebuttal_cid)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)

    def finalize_challenge(self, completion_id: int) -> dict:
        """Anyone finalizes an uncontested challenge after the window (upheld)."""
        fn = self.contract.functions.finalizeChallenge(completion_id)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 200_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)


class CapabilityRegistryClient:
    """Client for the CapabilityRegistry (LoRA Guilds) contract."""

    def __init__(self, address: str, w3: Web3, account: Optional[Account] = None):
        self.w3 = w3
        self.account = account
        self.contract = w3.eth.contract(address=address, abi=CAPABILITY_REGISTRY_ABI)

    @property
    def address(self) -> str:
        return self.contract.address

    def total_supply(self) -> int:
        return self.contract.functions.totalSupply().call()

    def token_by_index(self, index: int) -> int:
        return self.contract.functions.tokenByIndex(index).call()

    def get_capability(self, capability_id: int) -> Capability:
        r = self.contract.functions.getCapability(capability_id).call()
        return Capability(
            creator=r[0], bond=r[1], capability_type=r[2], metadata_cid=r[3],
            certified=r[4], slashed=r[5],
        )

    def get_capabilities_by_type(self, capability_type: str) -> list[int]:
        """v0.1.2: indexed lookup (O(1) per type) instead of a full scan."""
        return list(self.contract.functions.getCapabilitiesByType(_keccak(capability_type)).call())

    def count_capabilities_by_type(self, capability_type: str) -> int:
        """F10: number of live capabilities of a type."""
        return self.contract.functions.countCapabilitiesByType(_keccak(capability_type)).call()

    def get_capabilities_by_type_paged(self, capability_type: str, offset: int, limit: int) -> list[int]:
        """F10: paginated discovery view (v0.2 contracts)."""
        return list(
            self.contract.functions.getCapabilitiesByTypePaged(_keccak(capability_type), offset, limit).call()
        )

    def _send_tx(self, tx):
        if self.account is None:
            raise ValueError("No account set — cannot send transactions")
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        return self.w3.eth.wait_for_transaction_receipt(tx_hash)

    def register_capability_eth(self, capability_type: str, metadata_cid: str, bond_wei: int) -> dict:
        """Register a capability with an ETH bond. Returns {'capabilityId': int, 'receipt': dict}."""
        fn = self.contract.functions.registerCapabilityEth(_keccak(capability_type), metadata_cid)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "value": bond_wei,
            "gas": 300_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        receipt = self._send_tx(tx)
        # v0.1.2: derive the id from the emitted event. totalSupply() returns the
        # wrong id once any capability has been withdrawn (burn decrements supply,
        # ids keep incrementing).
        capability_id = _event_id_from_receipt(self.contract, receipt, "CapabilityRegistered", "capabilityId")
        if capability_id is None:
            capability_id = self.contract.functions.totalSupply().call()
        return {"capabilityId": capability_id, "receipt": receipt}

    def certify_capability(self, capability_id: int) -> dict:
        fn = self.contract.functions.certifyCapability(capability_id)
        tx = fn.build_transaction({
            "from": self.account.address,
            "nonce": self.w3.eth.get_transaction_count(self.account.address),
            "gas": 100_000,
            "gasPrice": self.w3.eth.gas_price,
            "chainId": self.w3.eth.chain_id,
        })
        return self._send_tx(tx)


def connect(rpc_url: str, chain_id: int) -> Web3:
    """Create a Web3 instance with POA middleware (needed for Base)."""
    w3 = Web3(Web3.HTTPProvider(rpc_url))
    if chain_id in (84532, 8453, 31337):
        # Base / Base Sepolia / hardhat may need POA middleware
        try:
            w3.middleware_onion.inject(ExtraDataToPOAMiddleware, layer=0)
        except Exception:
            pass  # Already injected or not needed
    return w3


def load_account(private_key: str) -> Account:
    """Load an eth-account from a private key string."""
    return Account.from_key(private_key)


def discover(
    registry: CapabilityRegistryClient,
    ron: ReputationOracleNetworkClient,
    capability_type: str = LORA_CAPABILITY_TYPE,
    min_score: int = 0,
) -> list[dict]:
    """Discover agents by capability proof + self-attest score. Mirrors /api/discover.

    v0.1.2: uses the indexed getCapabilitiesByType lookup and skips stale ids
    instead of failing the whole call when one id no longer resolves."""
    try:
        ids = registry.get_capabilities_by_type(capability_type)
    except Exception:
        total = registry.total_supply()
        ids = [registry.token_by_index(i) for i in range(total)]

    results = []
    for cap_id in ids:
        try:
            cap = registry.get_capability(cap_id)
        except Exception:
            continue  # stale/unreadable id — skip, never fail the whole discovery
        if not cap.certified or cap.slashed:
            continue
        # v0.2: rank on the receipt-confirmed (two-sided) score where supported.
        score, count, disputes, score_type = ron.get_ranking_score(cap.creator)
        if score < min_score:
            continue
        results.append({
            "agentAddress": cap.creator,
            "capabilityId": cap_id,
            "capabilityType": capability_type,
            "certified": cap.certified,
            "slashed": cap.slashed,
            "bond": cap.bond,
            "metadataCID": cap.metadata_cid,
            "completions": count,
            "disputes": disputes,
            "score": score,
            "scoreType": score_type,
        })
    results.sort(key=lambda x: x["score"], reverse=True)
    return results
